pupil_src/shared_modules/calibration_routines/calibrate_2d.py

```diff
@@ -0,0 +1,264 @@
+"""
+(*)~---------------------------------------------------------------------------
+Pupil - eye tracking platform
+Copyright (C) 2012-2018 Pupil Labs
+
+Distributed under the terms of the GNU
+Lesser General Public License (LGPL v3.0).
+See COPYING and COPYING.LESSER for license details.
+---------------------------------------------------------------------------~(*)
+"""
+
+import numpy as np
+import numpy.polynomial.polynomial as poly
+import scipy.optimize
+
+# logging
+import logging
+
+logger = logging.getLogger(__name__)
+
+
+class PolynomialMonocular:
+    """
+    A polynomial function class based on the polyval functions from numpy.polynomial.polynomial.
+
+    Currently supports 1d, 2d, and 3d polynomials.
+    """
+
+    def __init__(self, params, dof=2, degree=2):
+
+        self.dof = dof
+        self.degree = degree
+        self.func = [poly.polyval, poly.polyval2d, poly.polyval3d][self.dof - 1]
+        self.params = np.reshape(params, self.dof * [degree + 1])
+
+    def __call__(self, x):
+
+        return self.func(*x, self.params)
+
+
+class PolynomialBinocular:
+    """
+    The average of two monocular polynomials with independent parameters.
+    """
+
+    def __init__(self, params, dof=2, degree=2):
+
+        self.dof = dof
+        self.degree = degree
+
+        self.mapping_eye_0 = PolynomialMonocular(
+            params[: len(params) // 2], self.dof, self.degree
+        )
+        self.mapping_eye_1 = PolynomialMonocular(
+            params[len(params) // 2 :], self.dof, self.degree
+        )
+
+    def __call__(self, x):
+
+        x0 = x[: self.dof]
+        x1 = x[self.dof :]
+
+        return 1 / 2 * (self.mapping_eye_0(x0) + self.mapping_eye_1(x1))
+
+
+class MappingFunction2D:
+    """
+    Maps 2d pupil positions to 2d screen coordinates, using either monocular or binocular polynomials.
+    """
+
+    def __init__(self, params_x, params_y, dof=2, degree=2, binocular=False):
+
+        if binocular:
+            polynomial_type = PolynomialBinocular
+        else:
+            polynomial_type = PolynomialMonocular
+
+        self.map_to_world_cam_x = polynomial_type(params_x, dof, degree)
+        self.map_to_world_cam_y = polynomial_type(params_y, dof, degree)
+
+    def __call__(self, pt):
```

Comment:
I am not sure about this, so @papr correct me if I am wrong, but if all your classes only implement `__call__` and nothing else it might make more sense to use a factory instead. A factory is a function that takes parameters and returns a function, which is specified by these parameters. For example:
```
def add_factory(b):
    def add(a):
        return a+b
    return add
```

The difference between the function returned by the factory and the `__call__` function of your class is that the parameters that define what happens in `__call__` are technically still mutable, while the function returned by the factory is not. In your case you do not want the function to be mutable, so the factory pattern might be more appropriate.